stl/inc/type_traits

```diff
@@ -366,6 +366,18 @@ struct is_enum : bool_constant<__is_enum(_Ty)> {}; // determine whether _Ty is a
 template <class _Ty>
 _INLINE_VAR constexpr bool is_enum_v = __is_enum(_Ty);
 
+#if _HAS_CXX23
+// STRUCT TEMPLATE is_scoped_enum
+template <class _Ty, bool = conjunction_v<is_enum<_Ty>, negation<is_convertible<_Ty, int>>>>
+inline constexpr bool is_scoped_enum_v = true;
+
+template <class _Ty>
+inline constexpr bool is_scoped_enum_v<_Ty, false> = false;
+
+template <class _Ty>
+struct is_scoped_enum : bool_constant<is_scoped_enum_v<_Ty>> {};

```

Comment:
I am a bit torn, usually we always follow the convention to define the variable template and then derive the struct from it. 

In this case we require the short circuiting from `conjunction`, so maybe the "best" implementation would be to do
```suggestion
struct is_scoped_enum : conjunction<is_scoped_enum_v<_Ty>> {};
```

And then derive the variable template from that?

Reply:
This was originally under `is_unbounded_array` as that's where `is_scoped_enum` is depicted in the standard, and that implementation doesn't do that. (It was moved as is_enum and is_convertible are defined below).

Is there a general rule about this?

Reply:
I figure most people would be using the suffix version and doing it the traditional way would mean an extra template instantiation, wouldn't it?

Reply:
Our modern convention is to follow the Standard order when it's easy to do so - however, the type traits are fairly jumbled, partly due to history, partly due to their dependencies. I wouldn't worry about the ordering.

As for whether the struct or the variable template should be "fundamental" - it's the struct that's costly for throughput, whereas variable templates are cheap. The compiler has built-in optimizations for `conjunction_v`, `negation`, `is_convertible`, and I think `is_enum` too. So, the code should have good throughput as-is.

While it's possible to duplicate the implementation to improve the struct throughput slightly, as @miscco depicted, I don't think that's necessary here (because I don't think that this type trait will be used in a massive number of specializations - whereas the traits like `is_convertible` are highly used). Additionally, note that if we did duplicate the implementation, we would want to "smash" the base class to `true_type` or `false_type` - i.e. derive from `conjunction<ARGS>::type` (or `bool_constant<conjunction_v<ARGS>>`) and not `conjunction<ARGS>` directly. (If you search, you'll see a few occurrences of this throughput the STL.) This simplifies the actual inheritance.